lib/Target/ExportSystemC/EmissionPattern.h

```diff
@@ -0,0 +1,362 @@
+//===- EmissionPattern.h - Emission Pattern Base and Utility --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This declares the emission pattern base and utility classes.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef EMISSION_PATTERN_H
+#define EMISSION_PATTERN_H
+
+#include "EmissionPatternSupport.h"
+#include "mlir/IR/Operation.h"
+#include <any>
+
+namespace circt {
+namespace ExportSystemC {
+// Forward declarations.
+class EmissionPrinter;
+
+//===----------------------------------------------------------------------===//
+// Utilities to configure the behavior of emission patterns.
+//===----------------------------------------------------------------------===//
+
+/// A Flag to be added to an EmissionConfig.
+template <typename T>
+class Flag {
+public:
+  Flag<T>(StringRef name, T defaultValue)
+      : name(name), defaultValue(defaultValue) {}
+
+  StringRef getName() const { return name; }
+  T getDefault() const { return defaultValue; }
+
+private:
+  StringRef name;
+  T defaultValue;
+};
+
+/// This class collects a set of configuration flags to be passed to the
+/// emission patterns. Emission patterns can make their matching decision
+/// dependent on the configuration, or change their emission behavior.
+class EmissionConfig {
+public:
+  template <typename T>
+  void set(StringRef flag, T value) {
+    flags[flag] = value;
+  }
+
+  template <typename T>
+  T get(Flag<T> flag) {
+    if (!flags.count(flag.getName()))
+      return flag.getDefault();
+    return std::any_cast<T>(flags[flag.getName()]);
+  }
+
+private:
+  DenseMap<StringRef, std::any> flags;
+};
```

Comment:
Is there any reason that flags are defined dynamically? Can we define `EmissionConfig` similar to `LoweringOptions`?

Reply:
I completely agree with that. Thanks for linking the discussion. I will remove the emission config in this PR and add one based on attributes in a separate one, when we actually have a use-case.